typing/ctype.ml

```diff
@@ -2246,31 +2276,46 @@ let unbox_once env ty =
     begin match Env.find_type p env with
     | exception Not_found -> Missing p
     | decl ->
-      let apply ty2 = apply env decl.type_params ty2 args in
+      let apply ty2 ~extra_substs =
+        let extra_params, extra_args = List.split extra_substs in
+        (* put extras first as they're often empty. *)
+        apply env (extra_params @ decl.type_params) ty2 (extra_args @ args)
+      in
       begin match find_unboxed_type decl with
       | Some (ty2, modality) ->
         let ty2 = match get_desc ty2 with Tpoly (t, _) -> t | _ -> ty2 in
+        (* We need to ensure that existential variables do not escape their
+           scope. To do so, we substitute them with [Tof_kind]s. *)
         let existentials =
           match Env.find_type_descrs p env with
           | Type_variant ([{ cstr_existentials }], _, _) -> cstr_existentials
           | Type_variant (_not_one, _, _) ->
             Misc.fatal_error "Ctype.unbox_once: not just one constructor"
           | Type_abstract _ | Type_record _
           | Type_record_unboxed_product _ | Type_open -> []
-          | exception Not_found -> (* but we found it earlier! *) assert false
+          | exception Not_found ->
+            (* but we found it earlier! *)
+            Misc.fatal_error "Ctype.unbox_once: expected to find [p] in [env]"
         in
-        Stepped { ty = apply ty2;
-                  is_open = not (Misc.Stdlib.List.is_empty existentials);
-                  modality }
+        let extra_substs =
+          List.map
+            (fun ty ->
+              match get_desc ty with
+              | Tvar { name = _; jkind } -> ty, newgenty (Tof_kind jkind)
```

Comment:
Similar to my other comment, I'm not sure whether this should be `newty` rather than `newgenty`.